lib/services/role.go

```diff
@@ -723,6 +735,13 @@ type AccessChecker interface {
 	// "assume" while searching for resources, and should be able to request with a
 	// search-based access request.
 	GetSearchAsRoles() []string
+
+	// CanCreateHostUser determines whether a user is allowed to have a
+	// temporary user provisioned for them on a node
+	CanCreateHostUser(s types.Server) bool
+
+	// HostGroups returns all host groups matching a server.
+	HostGroups(s types.Server) []string
```

Comment:
These both have to iterate over roles, calculating a match against the same server, and they're both called at the same time; can we merge them so we only have to iterate once? Once we add sudoers support it's going to be a third iteration over the same thing, calculating the same things.